optaplanner-core/src/main/java/org/optaplanner/core/impl/domain/solution/util/ConcurrentCache.java

```diff
@@ -0,0 +1,31 @@
+package org.optaplanner.core.impl.domain.solution.util;
+
+import java.util.concurrent.ConcurrentHashMap;
+import java.util.function.Function;
+
+/**
+ * Thread-safe basic cache implementation.
+ *
+ * @param <K> the type of keys for the cache
+ * @param <V> the type of values stored in the cache
+ * */
+public final class ConcurrentCache<K, V> extends ConcurrentHashMap<K, V> {
+
+    /**
+     * {@link ConcurrentHashMap#computeIfAbsent(Object, Function)} should not be used to directly retrieve cached values
+     * as it incorporates locking even when the value is present and doesn't have to be computed. Thus, the value is
+     * retrieved by {@link ConcurrentHashMap#get(Object)} and if and only if the value is null,
+     * {@link ConcurrentHashMap#computeIfAbsent(Object, Function)} is used to ensure thread-safety.
```

Comment:
A good javadoc shouldn't discuss implementation details, just behavior guaranties towards the client.
The reason for this, if we want to change the code of this method, we'd want to avoid having to update the javadoc as that will be strictly-speaking a backwards incompatible change (yes this in not a public api, but still it is a bc breakage against all our internal classes using this, which is fixable of course, unlike public api bc breakage).

Anyway, I'd shorten this paragraph to something like:
"A computeIfAbsent() implementation that heavily favors read access over write access."